### ISSUE TITLE
No attribute 'asn_table' error when running cube builder

### ISSUE THREAD

**Issue by mcara** [STScI-JWST/jwst#988]
```python
>>> from jwst.datamodels import ModelContainer, ImageModel
>>> from jwst import cube_build
>>> cbs = cube_build.CubeBuildStep()
2017-06-20 13:46:36,222 - stpipe.CubeBuildStep - INFO - CubeBuildStep instance created.
>>> cbs.channel = '1'
>>> cbs.band = 'ALL'
>>> cbs.single = True      
>>> im = ImageModel('/grp/jwst/ssb/mcara/from_jane/MIRM103-Q0-SHORT_495_rate_bsub_cal.fits')
2017-06-20 13:49:04,089 - stpipe - WARNING - /.../lib/python2.7/site-packages/asdf/asdftypes.py:260: UserWarning: 'tag:stsci.edu:jwst_pipeline/v23tosky' with version 0.1.0 found in file, but asdf only understands version 0.7.0.
  warnings.warn(warning_string)

>>> models = ModelContainer([im])
>>> cbs.process(models)
2017-06-20 13:49:38,279 - stpipe.CubeBuildStep - INFO - Starting IFU Cube Building Step
2017-06-20 13:49:38,279 - stpipe.CubeBuildStep - INFO - Input interpolation: pointcloud
2017-06-20 13:49:38,279 - stpipe.CubeBuildStep - INFO - Coordinate system to use: ra-dec
2017-06-20 13:49:38,280 - stpipe.CubeBuildStep - INFO - Weighting method for point cloud: standard
2017-06-20 13:49:38,280 - stpipe.CubeBuildStep - INFO - Power Weighting distance : 2.000000
2017-06-20 13:49:38,280 - stpipe.CubeBuildStep - INFO -  Single = true, creating a set of single exposures mapped to output IFUCube coordinate system
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/.../jwst/cube_build/cube_build_step.py", line 132, in process
    input_table = data_types.DataTypes(input)
  File "/.../jwst/cube_build/data_types.py", line 69, in __init__
    self.output_name =input_model.meta.asn_table.products[0].name
  File "/.../jwst/datamodels/properties.py", line 205, in __getattr__
    raise AttributeError("No attribute '{0}'".format(attr))
AttributeError: No attribute 'asn_table'
>>> 
```

The data file can be obtained from ``/grp/jwst/ssb/mcara/from_jane/``

**Comment by mcara** [comment]
CC: @jemorrison @jdavies-st

**Comment by jemorrison** [comment]
Darn it
I am on it.
Jane

**Comment by hbushouse** [comment]
This may not be a problem with cube_build itself. When a ModelContainer is constructed automatically from an ASN, it populates the attribute 'asn_table' with meta info from the ASN file. When you construct a ModelContainer manually, as is being done here, from either a single data model or multiple data models, the asn_table attribute of the ModelContainer instance is not being populated. Do the inner workings of cube_build need the info in the asn_table attribute in this case? If so, then it will need to be populated before the ModelContainer is passed to cube_build. If not, then this particular branch of cube_build should be updated to not try to access it.

**Comment by mcara** [comment]
What is 'asn_table' and what it is for? Why does it need to be populated for a single file? @jemorrison  Why does cube builder need to know asn_table? Something is fishy here...

**Comment by mcara** [comment]
Also, without understanding what 'asn_table' is for but suspecting that it shows the ASN table from which the file name came, it seems natural that 'asn_table' be either ``None`` or not exist for stand-alone files (=not opened through an ASN table).

**Comment by jemorrison** [comment]
Ok I have limited internet - I think the lines have melted here because it
is only 113.
I can only read email on gmail - not my university account

I know what is happening in Mihia's case. The way  I understood how this
would work is an upper wrapper program calspec3 would open an association
table containing all the data and pass the data to each part of the
pipeline. When I have tested this I have a routine open
and association table, read the data into a model container and it also
reads in the association output name, with container =
datamodels.open(association_name).

At this stage in cube_build it does not care if single is set to true.
I can change this and if single is set to true then it does not try ad grab
the association
output name.

Jane

**Comment by mcara** [comment]
> At this stage in cube_build it does not care if single is set to true.

I think this is how it should be... Why would you treat input differently as a function of ``single`` argument? In my opinion the code should be designed to work with ModelContainer input that did not "originate" from an ASN.

**Comment by hbushouse** [comment]
Most of the time cube_build is likely to be presented with a ModelContainer that was built from an ASN table, but there are cases even within the calspec3 pipeline where that might be true. For example, a while back we had discussed the possibility that Mihai's skymatch step might build it's own ModelContainers that would include only the channel 1+2 exposures or only the channel 3+4 exposures, so that he can do an easier mapping of items in the input and output lists. If he constructs ModelContainers in that way, then they won't have the asn_table info available for use. So yes it would be best if cube_build did not rely on getting info from there, at least not in this case.

I've always thought that having datamodels.open() secretly populate the asn_table attribute with info from the ASN table, when it's given an ASN table as input, was a bit dangerous and asking for trouble. The only pieces of info you should expect to find in a ModelContainer is the list of individual data models. If you need other ancillary info, such as the product names, exptypes, etc. from the ASN table, then the ASN table itself should be explicitly loaded into its own object, using something like load_asn(), rather than having it tucked away in a special cubby within the ModelContainer object.

**Comment by jemorrison** [comment]
Ok right now checking if a ModelContainer was past in.
isinstance(input,datamodels.ModelContainer)

if it is then I have been filling in the output_name by  the hidden 
functions that datamodels.open does
Which is the problem for Mihai's case.

So just so we are all on the same page - should I check if single = true 
and skip reading in the output filename

But if single is not set - then if seems I probably should not be using 
the hidden routines of datamodels.open
and I should just use load_asn..

So my question is - if cube build is past a ModelContainer and single = 
false is it always the case that
I should load_asn ? But will load_asn work if it is a ModelContainer ? 
That does not make sense to me.

Jane

**Comment by hbushouse** [comment]
I agree that if "single" is set, there's no point in trying to access the asn_table attribute to get the output product name, because in this case there will be multiple output products and hence that single product name from asn_table isn't of any use.

When single is not set, however, there's still no guarantee that the input ModelContainer was originally created by loading the entire contents of an ASN file (via datamodels.open(asn_file)), so even if you want to use info from within asn_table you should still first check to make sure it exists and if it doesn't exist, then take some other suitable action. That's better than having the code crash when trying to access something from the non-existent asn_table attribute. And in the case where the original input is still in the form of an ASN file name, instead of doing just a datamodels.open() on it, also do an explicit load_asn() in order to get all the info from the ASN file that you need.

**Comment by jemorrison** [comment]
closed with #990
